### ISSUE TITLE
load_rect exception : if deprecating what is the alternative?

### ISSUE THREAD

**Issue by jramapuram** [PistonDevelopers/image#802]
I'm trying to use `image` + `rayon` to parallely crop a minibatch of images.
I'm trying to operate over medical images that are pretty large, thus the lazy load instead of `image::open` call as I won't be able to load more than 1-5 images otherwise.

Here is a simple crop function that tries to lazy load and crop an image:

```rust
pub fn lazy_crop_to_vec(path_str: &str, x: u32, y: u32, width: u32, length: u32) -> ImageResult<Vec<u8>>
{
    let path = Path::new(&path_str);
    let format = get_image_format(path).unwrap();

    // the file input reader
    let fin = match File::open(path) {
        Ok(f)  => f,
        Err(err) => return Err(image::ImageError::IoError(err))
    };
    let fin = BufReader::new(fin);

    println!("img_dims = {:?}, x = {:?} | y = {:?} | width = {:?} | len = {:?}", dimensions(&path_str).unwrap(), x, y, width, length);
    match format {
        image::ImageFormat::PNG  => png::PNGDecoder::new(fin).load_rect(x, y, length, width),
        image::ImageFormat::GIF  => gif::Decoder::new(fin).load_rect(x, y, length, width),
        image::ImageFormat::JPEG => jpeg::JPEGDecoder::new(fin).load_rect(x, y, length, width),
        image::ImageFormat::TIFF => try!(tiff::TIFFDecoder::new(fin)).load_rect(x, y, length, width),
        image::ImageFormat::TGA => tga::TGADecoder::new(fin).load_rect(x, y, length, width),
        image::ImageFormat::BMP => bmp::BMPDecoder::new(fin).load_rect(x, y, length, width),
        image::ImageFormat::ICO => try!(ico::ICODecoder::new(fin)).load_rect(x, y, length, width),
        image::ImageFormat::HDR => try!(hdr::HDRAdapter::new(BufReader::new(fin))).load_rect(x, y, length, width),
        image::ImageFormat::PNM => try!(pnm::PNMDecoder::new(BufReader::new(fin))).load_rect(x, y, length, width),
        _ => Err(image::ImageError::UnsupportedError(format!("A decoder for {:?} is not available.", format))),
    }
}
```

however, it fails due to the following exception on a PNG image even though the provided bounds are valid, i.e. 	***img_dims = (512, 512), x = 256 | y = 256 | width = 128 | len = 128*** :

```bash
(base) ➜  parallel_image_crop git:(master) ✗ cargo test --lib
    Blocking waiting for file lock on build directory
    Finished dev [unoptimized + debuginfo] target(s) in 0.78s
     Running target/debug/deps/parallel_image_crop-337e6d61909c67aa

running 2 tests
test tests::test_bw_image_crop ... FAILED
test tests::test_image_crops ... FAILED

failures:

---- tests::test_bw_image_crop stdout ----
	img_dims = (512, 512), x = 256 | y = 256 | width = 128 | len = 128
thread 'tests::test_bw_image_crop' panicked at 'called `Result::unwrap()` on an `Err` value: ImageEnd', libcore/result.rs:945:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Looks like the following line is being triggered in `read_scanline` of `src/png.rs` in `image`:

```rust
None => Err(ImageError::ImageEnd)
```

If `load_rect` is being deprecated, what is the alternative to not loading the entire image?

**Comment by jramapuram** [comment]
Would also like to note that using `image::open( ... ).crop(x, y, width, length)` works fine.

**Comment by fintelia** [comment]
I've made an alternate proposal in #798. In that design, load_rect should never fail, but for certain image types might allocate

**Comment by fintelia** [comment]
I think this is actually caused by #709. Specifically, load_rect keeps reading rows from the image until it gets to the ones that make up the rectangle, but read_scanline erroneously returns the row lengths instead of indices. As a result, load_rect runs off the end of the image and returns the error you are seeing.

**Comment by fintelia** [comment]
Regarding the future of load_rect, I've written up a [proposal to restructure the ImageDecoder trait](#798) which load_rect is part of. Any feedback is welcome

**Comment by jramapuram** [comment]
@fintelia : I will try the same op with `jpeg` as a trial and get back to you. However, as mentioned in another issue here:

 Looking through the code though, I see that `JPEGDecoder` and `PNMDecoder` don't implement `read_scanline` ?

```rust
impl<R: Read> ImageDecoder for JPEGDecoder<R> {
...
    fn read_scanline(&mut self, _buf: &mut [u8]) -> ImageResult<u32> {
        unimplemented!();
    }
}
```
It also looks like all the `load_rect` calls use `read_scanline` ? 
```rust
        loop {
            let row = try!(self.read_scanline(&mut tmp));

            if row - 1 == y {
                break
            }
        }
```

So technically I can't use `load_rect` for `jpeg`'s at this time. 

Will take a look at the RFC though, thanks!

**Comment by fintelia** [comment]
The proposed changes I mentioned earlier are now merged via #819
